libraries/AP_UAVCAN_V1/AP_UAVCAN_V1.cpp

```diff
@@ -0,0 +1,405 @@
+    // @Range: 0 6143
+    // @User: Advanced
+    AP_GROUPINFO("ST4",     23, AP_UAVCAN_V1, _parameters_table[22], 0),
+
+    // @Param: DYN4
+    // @DisplayName: Dynamics subject id
+    // @Description: reg.udral.physics.dynamics.rotation.PlanarTs_0_1
+    // @Range: 0 6143
+    // @User: Advanced
+    AP_GROUPINFO("DYN4",    24, AP_UAVCAN_V1, _parameters_table[23], 0),
+
+    AP_GROUPEND
+};
+
+
+///< Base must be aligned for O1HEAP_ALIGNMENT byes. Othwerwise o1heapInit may fails!
+uint8_t base[UAVCAN_HEAP_SIZE] __attribute__ ((aligned (O1HEAP_ALIGNMENT)));
+
+static O1HeapInstance* my_allocator;
+
+
+static void* memAllocate(CanardInstance* const canard_, const size_t amount);
+static void memFree(CanardInstance* const canard_, void* const pointer);
+
+
+void* memAllocate(CanardInstance* const canard_, const size_t amount)
+{
+    (void) canard_;
+    return o1heapAllocate(my_allocator, amount);
+}
+void memFree(CanardInstance* const canard_, void* const pointer)
+{
+    (void) canard_;
+    o1heapFree(my_allocator, pointer);
+}
+
+
+
+AP_UAVCAN_V1 *AP_UAVCAN_V1::get_uavcan(uint8_t driver_index)
+{
+    if (driver_index >= AP::can().get_num_drivers() ||
+        AP::can().get_driver_type(driver_index) != AP_CANManager::Driver_Type_UAVCAN_V1) {
+        return nullptr;
+    }
+    return static_cast<AP_UAVCAN_V1*>(AP::can().get_driver(driver_index));
+}
+
+void AP_UAVCAN_V1::init(uint8_t driver_index, bool enable_filters)
+{
+    if (_initialized) {
+        return;
+    }
+
+    my_allocator = o1heapInit(base, UAVCAN_HEAP_SIZE);
+    if (NULL == my_allocator) {
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "v1: o1heapInit() failed");
+        return;
+    }
+
+    _canard = canardInit(&memAllocate, &memFree);
+    _canard.node_id = 42;
+    _tx_queue = canardTxInit(UAVCAN_TX_QUEUE_FRAME_SIZE, CANARD_MTU_CAN_CLASSIC);
+
+    publisher_manager.init(_canard, _tx_queue);
+    subscriber_manager.init(_canard, _tx_queue);
+
+    _registers.init(subscriber_manager, _canard, _tx_queue);
+    _esc_controller.init(subscriber_manager, publisher_manager, _canard, _tx_queue);
+
+    bool is_thread_created = hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_UAVCAN_V1::loop, void),
+                             _thread_name,
+                             UAVCAN_STACK_SIZE,
+                             AP_HAL::Scheduler::PRIORITY_CAN,
+                             0);
+    if (!is_thread_created) {
+        return;
+    }
+
+    _initialized = true;
+}
+
+void AP_UAVCAN_V1::loop(void)
+{
+    while (true) {
+        hal.scheduler->delay_microseconds(500); // it should be enough to handle max frame rate
+
+        // do nothing until interface initialization is complete
+        if (_can_iface == nullptr) {
+            continue;
+        }
+
+        spinReceive();
+        spinTransmit();
+        publisher_manager.process_all();
+
+        static uint32_t next_pub_time_ts = 1000;
+        if (AP_HAL::millis() > next_pub_time_ts) {
```

Comment:
why publish once a seconds ? just publish when you get a new message

Reply:
fixed